relay/environment/environment.go

```diff
@@ -0,0 +1,214 @@
+func (env *Map) LookupRequired(key string) (string, error) {
+	val, ok := env.provider.Lookup(key)
+	if ok {
+		return val, nil
+	}
+	return "", fmt.Errorf("Missing required configuration variable: %v", key)
+}
+
+// SubstituteVars substitutes variables from the environment into the provided
+// YAML string. "${FOO}" is replaced with the value of the variable "FOO".
+// "${FOO:BAR}" is also replaced by the value of "FOO", but if "FOO" is not
+// present, the value "BAR" is used instead. If "$(FOO)" or "$(FOO:BAR")" is
+// used, no escaping is performed before substitution.
+func (env *Map) SubstituteVarsIntoYaml(input string) string {
+	return varSubstitutionRegexp.ReplaceAllStringFunc(input, func(expression string) string {
+		submatches := varSubstitutionRegexp.FindStringSubmatch(expression)
+
+		var envVar string
+		var defaultVal string
+		var escape func(input string) string
+
+		if submatches[1] != "" {
+			// We've got ${VAR} or ${VAR:DEFAULT}.
+			envVar = submatches[2]
+			defaultVal = submatches[4]
+
+			// For this kind of substitution, we attempt to autodetect the
+			// appropriate YAML type for the value and transform it. Note that
+			// the result is always a "primitive" YAML value, and never an
+			// arbitrary hunk of YAML syntax.
+			escape = func(value string) string {
+				// Leave values that are some kind of non-string YAML primitive
+				// unchanged.
+				if nullValueRegexp.MatchString(value) ||
+					boolValueRegexp.MatchString(value) ||
+					intValueRegexp.MatchString(value) ||
+					floatValueRegexp.MatchString(value) {
+					return value
+				}
+
+				// Default to treating this value as a string. We pass it
+				// through yaml.Marshal() to ensure that it's correctly escaped.
+				if yamlBytes, err := yaml.Marshal(&value); err == nil {
+					// yaml.Marshal() will insert a newline after the literal
+					// value it generates, so we need to remove it.
+					return strings.TrimRight(string(yamlBytes), "\n")
+				}
+
+				// The input is invalid; just return the empty string.
+				logger.Printf(`Invalid value for environment variable '%v': %v`, envVar, value)
+				return ""
+			}
+		} else {
+			// We've got $(VAR) or $(VAR:DEFAULT).
+			envVar = submatches[6]
+			defaultVal = submatches[8]
+
+			// For this kind of substitution, we just substitute in the value
+			// directly, without any transformations. This is usually not
+			// desirable because it requires you to format your environment
+			// variables as a valid YAML value and deal with quotes and escape
+			// sequences, but it gives the user full control over the
+			// substitution process.
+			escape = func(value string) string {
+				return value
+			}
+		}
+
+		// As a special case, if the variable name is the empty string, just
+		// return the default value, which may also be empty. This provides a
+		// way to escape the literal string "${FOO}"; you can write it as
+		// "$${}{FOO}", and the "${}" will disappear after substitution.
```

Comment:
NIT: this seems like a really unusual way of having a literal value of `${FOO}`, could we have either added something that look for `(?:^|[^\\])` (start or a character that isn't `\`) to allow it to be escaped as `\${FOO}` in the actual value, or does it not work in go? there probably isn't a terribly often case of this but this still seems awkward

Reply:
I was a bit hesitant to do this initially because of the interaction with YAML's normal backslash escape rules. In YAML, a backslash is a literal backslash _unless_ it comes before an escapable character. A backslash is an escapable character, so `\\` is translated to `\` in YAML. A `$` is not, so `\$` is translated to `\$` in YAML - the backslash is literal and remains in the output. So, if we treat `\${FOO}` as `${FOO}`, we're not just escaping `${FOO}`; we're also removing a backslash that would normally show up literally in the output.

That said, I agree that this is probably the right move. To make it feel right, I think we need to count the total number of backslashes before the `$`. That way, we'll be able to understand if a backslash immediately before the `$` is escaped or not. We'll also be able to ensure that backslashes before the `$` always come in pairs, so a backslash before a `$` never ends up escaping the result of a substitution, which could be weird.

Here's an example of what I'm thinking the output should be. Assume you have an environment variable `FOO` with value `BAR`. I've listed three stages: (original text) -> (after our preprocessor) -> (after YAML parsing).
* `${FOO}` -> `BAR` -> `BAR`. (Zero backslashes; substitute.)
* `\${FOO}` -> `${FOO}` -> `${FOO}`. (Odd number of backslashes; remove one and don't substitute.)
* `\\${FOO}` -> `\\BAR` -> `\BAR`. (Even number of backslashes; leave as-is and substitute.)
* `\\\${FOO}` -> `\\${FOO}` -> `\${FOO}`. (Odd number of backslashes; remove one and don't substitute.)
* `\\\\${FOO}` -> `\\\\BAR` -> `\\BAR`. (Even number of backslashes; leave as-is and substitute.)

I'll make this change.